### ISSUE TITLE
snap not working

### ISSUE THREAD

**Issue by oSoMoN** [protonmail-desktop/application#170]
I've noticed the releases page has a snap package for amd64, which is great. I side-loaded it on Ubuntu 18.04 and the app starts but remains blocked on the splashscreen, I can't get any further.

I'm willing to contribute to the snap packaging but the snapcraft.yaml file is not in the repository, is that intended?

**Comment by kontrollanten** [comment]
Thanks for your feedback, @oSoMoN! 

Strange. I actually don't why it's not working. I know that there's issues when starting the app without network connection. May this be the case for you? I assume you've tried to restart the app?

We haven't created any snapcraft.yml. You're more then welcome to create one!

**Comment by popey** [comment]
@oSoMoN this is an electron-builder built application, so no snapcraft.yaml needed. The package.json has 'snap' as a listed output target, so when electron-builder is called during the application build, it then stands up a yaml and fires off snapcraft to build the snap. So the developer ends up with multiple build artifacts, including a snap, without having to make a yaml themselves.

**Comment by oSoMoN** [comment]
@popey thanks, I had just figured it out after reading the doc at https://docs.snapcraft.io/build-snaps/electron. Now looking into why the snap isn't working here.

**Comment by popey** [comment]
For me I get a popup telling me it can't find out if there's an app update. I then see this in the terminal where I launched the app.

```
$ protonmail-desktop 
Checking for update
Error: Error: ENOENT: no such file or directory, open '/snap/protonmail-desktop/x1/resources/app-update.yml'
    at /snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/src/AppUpdater.ts:362:27
    at Generator.next (<anonymous>)
From previous event:
    at AppImageUpdater.loadUpdateConfig (/snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/out/AppUpdater.js:380:11)
    at Lazy.AppUpdater.configOnDisk.Lazy [as creator] (/snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/src/AppUpdater.ts:125:43)
    at Lazy.get value [as value] (/snap/protonmail-desktop/x1/resources/app.asar/node_modules/lazy-val/src/main.ts:18:23)
    at /snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/src/AppUpdater.ts:273:33
    at Generator.next (<anonymous>)
From previous event:
    at AppImageUpdater.doCheckForUpdates (/snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/out/AppUpdater.js:338:11)
    at /snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/src/AppUpdater.ts:256:25
    at Generator.next (<anonymous>)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at AppImageUpdater._checkForUpdates (/snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/out/AppUpdater.js:292:11)
    at AppImageUpdater.checkForUpdates (/snap/protonmail-desktop/x1/resources/app.asar/node_modules/electron-updater/src/AppUpdater.ts:205:35)
    at App.app$1.on (/snap/protonmail-desktop/x1/resources/app.asar/app/src/auto-updater.js:10:17)
    at emitTwo (events.js:111:20)
    at App.emit (events.js:194:7)
```

**Comment by Lvaskz** [comment]
How can I install an application.Snap?  In the future we could download file.Snap  instead use the terminal? There is some plan about that?
I will ask snapcrafters too

**Comment by oSoMoN** [comment]
Ideally that snap would be published to the snap store, so it would be discoverable and installable on all distributions that support it.

**Comment by kontrollanten** [comment]
@oSoMoN That's a great idea. But first we need to make it working. I'm not sure if I've time to look at this problem this week.

@popey Would you have the ability to clone the repo, update electron-builder and see if it works?

**Comment by popey** [comment]
I tried to build it in a clean VM but I may be doing something wrong...

```
# yarn run dist
yarn run v1.5.1
$ electron-builder
  • electron-builder version=20.2.1
  • loaded configuration file=package.json ("build" field)
  • writing effective config file=dist/electron-builder.yaml
  • rebuilding native production dependencies platform=linux arch=x64
  • rebuilding native dependency name=node-sass
  • packaging       platform=linux arch=x64 electron=1.7.11 appOutDir=dist/linux-unpacked
Error: Application entry file "app/background.js" in the "/root/protonmail/application/dist/linux-unpacked/resources/app.asar" does not exist. Seems like a wrong configuration.
    at error (/root/protonmail/application/node_modules/electron-builder-lib/src/asar/asarFileChecker.ts:7:12)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/asar/asarFileChecker.ts:33:11
    at Generator.next (<anonymous>)
    at /root/protonmail/application/node_modules/graceful-fs/polyfills.js:287:18
    at FSReqWrap.oncomplete (fs.js:171:5)
From previous event:
    at checkFileInArchive (/root/protonmail/application/node_modules/electron-builder-lib/out/asar/asarFileChecker.js:46:21)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/platformPackager.ts:325:13
    at Generator.next (<anonymous>)
From previous event:
    at LinuxPackager.checkFileInPackage (/root/protonmail/application/node_modules/electron-builder-lib/out/platformPackager.js:398:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/platformPackager.ts:373:16
    at Generator.next (<anonymous>)
    at /root/protonmail/application/node_modules/graceful-fs/polyfills.js:287:18
    at FSReqWrap.oncomplete (fs.js:171:5)
From previous event:
    at LinuxPackager.sanityCheckPackage (/root/protonmail/application/node_modules/electron-builder-lib/out/platformPackager.js:416:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/platformPackager.ts:211:16
    at Generator.next (<anonymous>)
    at runCallback (timers.js:763:18)
    at tryOnImmediate (timers.js:734:5)
    at processImmediate (timers.js:716:5)
From previous event:
    at LinuxPackager.doPack (/root/protonmail/application/node_modules/electron-builder-lib/out/platformPackager.js:269:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/platformPackager.ts:116:16
    at Generator.next (<anonymous>)
From previous event:
    at LinuxPackager.pack (/root/protonmail/application/node_modules/electron-builder-lib/out/platformPackager.js:189:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/packager.ts:372:24
    at Generator.next (<anonymous>)
    at xfs.stat (/root/protonmail/application/node_modules/fs-extra/lib/mkdirs/mkdirs.js:56:16)
    at /root/protonmail/application/node_modules/graceful-fs/polyfills.js:287:18
    at FSReqWrap.oncomplete (fs.js:171:5)
From previous event:
    at Packager.doBuild (/root/protonmail/application/node_modules/electron-builder-lib/out/packager.js:441:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/packager.ts:316:52
    at Generator.next (<anonymous>)
    at /root/protonmail/application/node_modules/graceful-fs/graceful-fs.js:99:16
    at /root/protonmail/application/node_modules/graceful-fs/graceful-fs.js:43:10
    at FSReqWrap.oncomplete (fs.js:153:20)
From previous event:
    at Packager._build (/root/protonmail/application/node_modules/electron-builder-lib/out/packager.js:385:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/packager.ts:278:23
    at Generator.next (<anonymous>)
    at runCallback (timers.js:763:18)
    at tryOnImmediate (timers.js:734:5)
    at processImmediate (timers.js:716:5)
From previous event:
    at Packager.build (/root/protonmail/application/node_modules/electron-builder-lib/out/packager.js:341:11)
    at /root/protonmail/application/node_modules/electron-builder-lib/src/index.ts:50:40
    at Generator.next (<anonymous>)
From previous event:
    at build (/root/protonmail/application/node_modules/electron-builder-lib/out/index.js:47:21)
    at build (/root/protonmail/application/node_modules/electron-builder/src/builder.ts:228:10)
    at then (/root/protonmail/application/node_modules/electron-builder/src/cli/cli.ts:48:33)
    at runCallback (timers.js:763:18)
    at tryOnImmediate (timers.js:734:5)
    at processImmediate (timers.js:716:5)
From previous event:
    at Object.args [as handler] (/root/protonmail/application/node_modules/electron-builder/src/cli/cli.ts:48:33)
    at Object.runCommand (/root/protonmail/application/node_modules/yargs/lib/command.js:235:44)
    at Object.parseArgs [as _parseArgs] (/root/protonmail/application/node_modules/yargs/yargs.js:1042:24)
    at Object.get [as argv] (/root/protonmail/application/node_modules/yargs/yargs.js:957:21)
    at Object.<anonymous> (/root/protonmail/application/node_modules/electron-builder/src/cli/cli.ts:42:15)
    at Module._compile (module.js:662:30)
    at Object.Module._extensions..js (module.js:673:10)
    at Module.load (module.js:575:32)
    at tryModuleLoad (module.js:515:12)
    at Function.Module._load (module.js:507:3)
    at Function.Module.runMain (module.js:703:10)
    at startup (bootstrap_node.js:193:16)
    at bootstrap_node.js:660:3
```

**Comment by kontrollanten** [comment]
It seems that there has been some breaking changes in `electron-builder`. I'll take a look onto it and see if I can change what we need to fix.

**Comment by kontrollanten** [comment]
Sorry. It's late here. It seems that you didn't build the app before running `yarn dist`. Maybe that should be clarified. `yarn build` should create `app/background.js`.

**Comment by kontrollanten** [comment]
I've tried with latest electron-builder, but I get the same error as @popey regarding app-update.yml. According to https://www.electron.build/auto-update this file should be created by electron-builder. I don't have time to dig deeper into this just now though.

**Comment by kontrollanten** [comment]
@oSoMoN Can you try to download our latest pre-release to see if it works?

**Comment by kontrollanten** [comment]
I tried this in Ubuntu 16.04 and got it working by running

```
snap install --dangerous protonmail-desktop_1.0.0-beta.4_amd64.snap
protonmail-desktop
```

**Comment by flexiondotorg** [comment]
Hey @kontrollanten I work on the Snapcraft team and would love to help get ProtonMail Desktop published in the [Snap Store](https://snapcraft.io/store) :smiley: 

I'm a ProtonMail user myself, I've just tested the snap you have in your GitHub releases and it works a treat! I see you're using Circle CI, here is their documentation for publishing snaps in the store:

  * https://circleci.com/docs/2.0/build-publish-snap-packages/
  * https://circleci.com/blog/build-test-publish-snap-packages/

Once published, you'll have direct control of the publishing life cycle and users will get automatic updates. I'd love to feature ProtonMail Desktop as an editors pick and put together a social campaign to promote it. What additional help do you need to get ProtonMail Desktop published in the snap store?

**Comment by kontrollanten** [comment]
Hey @flexiondotorg ! That sounds amazing! I've had plans for snap store publishing myself, but I've had other stuff to prioritize.

To be the boring one. If the snap is working accurate, then I'll close this issue. Can you create a new issue where we discuss this further?

**Comment by oSoMoN** [comment]
@kontrollanten sorry for the late feedback, but I can confirm the snap for the 1.0.0-beta.7 release works well here. I too would very much welcome an officially supported snap in the store, rather than having to side-load a file I have download. Thanks!
